File: packages/core/src/ir.ts
// ASTronaut IR — TypeScript surface aligned with proto/parser.proto.
//
// These types follow the proto3 canonical JSON mapping so subprocess parsers
// (Go, .NET) can emit JSON that round-trips through here. The .proto file is
// the source of truth; if you change one, change both.

export type HttpMethod =
  | "GET"
  | "POST"
  | "PUT"
  | "DELETE"
  | "PATCH"
  | "HEAD"
  | "OPTIONS"
  | "TRACE";

export type SchemaKind =
  | "PRIMITIVE"
  | "OBJECT"
  | "ARRAY"
  | "REFERENCE"
  | "ONE_OF"
  | "ANY_OF"
  | "ALL_OF";

export type PrimitiveType = "string" | "integer" | "number" | "boolean" | "null";

export type ParamLocation = "header" | "query" | "cookie";

export type AuthType = "http" | "apiKey" | "oauth2" | "openIdConnect";

export type Severity = "warning" | "error";

export interface Constraints {
  minimum?: number;
  maximum?: number;
  exclusiveMinimum?: boolean;
  exclusiveMaximum?: boolean;
  minLength?: number;
  maxLength?: number;
  minItems?: number;
  maxItems?: number;
  uniqueItems?: boolean;
  pattern?: string;
  enumValues?: string[];
  multipleOf?: number;
}

export interface Schema {
  kind: SchemaKind;
  refName?: string;
  primitiveType?: PrimitiveType;
  format?: string;
  properties?: Record<string, Schema>;
  requiredProperties?: string[];
  items?: Schema;
  variants?: Schema[];
  constraints?: Constraints;
  nullable?: boolean;
  description?: string;
  defaultValue?: string;
  example?: string;
}

export interface ParamInfo {
  name: string;
  schema: Schema;
  required: boolean;
  description?: string;
  example?: string;
}

export interface BodyInfo {
  contentType: string;
  schema: Schema;
  required: boolean;
}

export interface ResponseInfo {
  status: number;
  description: string;
  schema?: Schema;
  contentType?: string;
  headers?: Record<string, ParamInfo>;
}

export interface AuthInfo {
  type: AuthType;
  scheme?: string;
  name?: string;
  in?: ParamLocation;
  bearerFormat?: string;
  id: string;
}

export interface SourceLocation {
  file: string;
  line: number;
  column: number;
}

export interface RouteInfo {
  method: HttpMethod;
  path: string;
  pathParams?: ParamInfo[];
  queryParams?: ParamInfo[];
  headerParams?: ParamInfo[];
  requestBody?: BodyInfo;
  responses?: ResponseInfo[];
  auth?: AuthInfo;
  summary?: string;
  description?: string;
  operationId?: string;
  tags?: string[];
  deprecated?: boolean;
  source: SourceLocation;
}

export interface ParseError {
  file: string;
  line: number;
  message: string;
  severity: Severity;
  code?: string;
}

export interface ParserMetadata {
  framework: string;
  frameworkVersion?: string;
  filesScanned: number;
  durationMs: number;
  parserVersion: string;
}

export interface ParseResult {
  routes: RouteInfo[];
  errors: ParseError[];
  metadata: ParserMetadata;
  sharedSchemas?: Record<string, Schema>;
}

File: analyzers/dotnet/src/AsTronaut.Analyzer/Ir/IR.cs
using System.Text.Json.Serialization;

namespace AsTronaut.Analyzer.Ir;

// IR DTOs — proto-aligned (camelCase JSON via JsonSerializerOptions). Fields use
// strings instead of C# enums for HttpMethod / SchemaKind / PrimitiveType / AuthType
// so that the JSON wire format exactly matches what the TS zod validator expects.

public sealed record ParseResult
{
    public List<RouteInfo> Routes { get; init; } = new();
    public List<ParseError> Errors { get; init; } = new();
    public ParserMetadata Metadata { get; init; } = new();

    [JsonPropertyName("sharedSchemas")]
    public Dictionary<string, Schema>? SharedSchemas { get; init; }
}

public sealed record RouteInfo
{
    // "GET" | "POST" | "PUT" | "DELETE" | "PATCH" | "HEAD" | "OPTIONS" | "TRACE"
    public string Method { get; init; } = "";
    public string Path { get; init; } = "";

    public List<ParamInfo>? PathParams { get; init; }
    public List<ParamInfo>? QueryParams { get; init; }
    public List<ParamInfo>? HeaderParams { get; init; }

    public BodyInfo? RequestBody { get; init; }
    public List<ResponseInfo>? Responses { get; init; }

    public AuthInfo? Auth { get; init; }

    public string? Summary { get; init; }
    public string? Description { get; init; }
    public string? OperationId { get; init; }
    public List<string>? Tags { get; init; }
    public bool? Deprecated { get; init; }

    public SourceLocation Source { get; init; } = new();
}

public sealed record ParamInfo
{
    public string Name { get; init; } = "";
    public Schema Schema { get; init; } = new();
    public bool Required { get; init; }
    public string? Description { get; init; }
    public string? Example { get; init; }
}

public sealed record BodyInfo
{
    public string ContentType { get; init; } = "application/json";
    public Schema Schema { get; init; } = new();
    public bool Required { get; init; }
    // All accepted content types (from [Consumes]); all share Schema.
    public List<string>? ContentTypes { get; init; }
}

public sealed record ResponseInfo
{
    public int Status { get; init; }
    public string Description { get; init; } = "";
    public Schema? Schema { get; init; }
    public string? ContentType { get; init; }
    public Dictionary<string, ParamInfo>? Headers { get; init; }
    // All produced content types (from [Produces]); all share Schema.
    public List<string>? ContentTypes { get; init; }
}

public sealed record Schema
{
    // "PRIMITIVE" | "OBJECT" | "ARRAY" | "REFERENCE" | "ONE_OF" | "ANY_OF" | "ALL_OF"
    public string Kind { get; init; } = "PRIMITIVE";

    public string? RefName { get; init; }

    // "string" | "integer" | "number" | "boolean" | "null"
    public string? PrimitiveType { get; init; }

    // OpenAPI format hint: "int32" | "int64" | "date-time" | "uuid" | "email" | ...
    public string? Format { get; init; }

    public Dictionary<string, Schema>? Properties { get; init; }
    public List<string>? RequiredProperties { get; init; }
    public Schema? Items { get; init; }
    public List<Schema>? Variants { get; init; }

    public Constraints? Constraints { get; init; }
    public bool? Nullable { get; init; }

    public string? Description { get; init; }
    public string? DefaultValue { get; init; }
    public string? Example { get; init; }
}

public sealed record Constraints
{
    public double? Minimum { get; set; }
    public double? Maximum { get; set; }
    public double? ExclusiveMinimum { get; set; }
    public double? ExclusiveMaximum { get; set; }
    public int? MinLength { get; set; }
    public int? MaxLength { get; set; }
    public int? MinItems { get; set; }
    public int? MaxItems { get; set; }
    public bool? UniqueItems { get; set; }
    public string? Pattern { get; set; }
    public List<string>? EnumValues { get; set; }
    public double? MultipleOf { get; set; }
}

public sealed record AuthInfo
{
    // "http" | "apiKey" | "mutualTLS" | "oauth2" | "openIdConnect"
    public string Type { get; init; } = "";
    public string? Scheme { get; init; }
    public string? Name { get; init; }
    // "header" | "query" | "cookie"
    public string? In { get; init; }
    public string? BearerFormat { get; init; }
    public string Id { get; init; } = "";
}

public sealed record ParseError
{
    public string File { get; init; } = "";
    public int Line { get; init; }
    public string Message { get; init; } = "";
    // "warning" | "error"
    public string Severity { get; init; } = "error";
    public string? Code { get; init; }
}

public sealed record ParserMetadata
{
    public string Framework { get; init; } = "aspnet";
    public string? FrameworkVersion { get; init; }
    public int FilesScanned { get; init; }
    public long DurationMs { get; init; }
    public string ParserVersion { get; init; } = "0.0.1";
}

public sealed record SourceLocation
{
    public string File { get; init; } = "";
    public int Line { get; init; }
    public int Column { get; init; }
}
